src/main/java/com/cryptomorin/xseries/particles/ParticleDisplay.java

```diff
@@ -1533,91 +1570,93 @@ public Location spawn(double x, double y, double z) {
+        double offsetx = offset.getX();
+        double offsety = offset.getY();
+        double offsetz = offset.getZ();
+        spawnRaw(particle, loc, count, offsetx, offsety, offsetz, data);
         return loc;
     }
 
-    private void updateColorType() {
-        if (data == null || !(data instanceof float[])) return;
-        float[] datas = (float[]) this.data;
-        if (this.particle == XParticle.NOTE) {
-            if (!isNoteColor) {
-                int nearestColor = findNearestNoteColor(new Color((int)datas[0], (int)datas[1], (int)datas[2]));
-                data = new float[]{nearestColor / 24f * 255, 0, 0, datas[3]};
-                isNoteColor = true;
-            }
-        } else if (isNoteColor) {
-            Color color = NOTE_COLORS[(int) (datas[0] / 255 * 24)];
-            data = new float[]{color.getRed(), color.getGreen(), color.getBlue(), datas[3]};
-            isNoteColor = false;
+    /**
+     * Spawns the particles with specific data in the offset fields.
+     * If required, this method will manually calculate an offset for
+     * each particle similarly to the standard behavior, and spawn the particles at those.
+     *
+     * @param offsetData the data that needs to go in the offset fields.
+     */
+    private void spawnWithDataInOffset(Particle particle, Location loc, double[] offsetData, Object data) {
+        // If there is no offset and we only want a single particle, we don't actually need to do anything special.
+        // Otherwise, we'll at least need to use a loop.
+        if (isZero(offset) && count < 2) {
+            spawnRaw(particle, loc, 0, offsetData[0], offsetData[1], offsetData[2], data);
+            return;
+        }
+        // Particles with a specific direction must be flagged with count = 0,
+        // so we have to spawn each particle manually.
+        double offsetx = offset.getX();
+        double offsety = offset.getY();
+        double offsetz = offset.getZ();
+        ThreadLocalRandom r = ThreadLocalRandom.current();
+        for (int i = 0; i < count; i++) {
+            // When specifying an offset normally, bound of 1 gets you an 8 block range,
+            // being +/- 4 blocks in each direction from the origin. Uses a Gaussian distribution.
+            // Gaussian distribution uses a sqrt, so skip that if we can.
+            double dx = offsetx == 0 ? 0 : r.nextGaussian() * 4 * offsetx;
+            double dy = offsety == 0 ? 0 : r.nextGaussian() * 4 * offsety;
+            double dz = offsetz == 0 ? 0 : r.nextGaussian() * 4 * offsetz;
+            Location offsetLoc = cloneLocation(loc).add(dx, dy, dz);
+            spawnRaw(particle, offsetLoc, 0, offsetData[0], offsetData[1], offsetData[2], data);
         }
     }
 
+    /**
+     * Calls the appropriate spawnParticle method with the parameters given.
+     */
+    private void spawnRaw(Particle particle, Location loc, int count, double dx, double dy, double dz, Object data) {
```

Comment:
Just use a `Vector` here for the offset then define dx,dy,dz variables for each inside the method since they're used 3 times.